pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,240 @@
+	//
+	// 1. The kernel doesn't support bpf_link and link.AttachCgroup attached a
+	// program to the hook using PROG_ATTACH with the multi flag. We need to make
+	// sure any other progs are gone.
+	//
+	// 2. The program was attached using bpf_link, and there was no preexisting
+	// pinned link. The link needs to be pinned and closed.
+	err = tryPin(l, name, pinPath)
+	if errors.Is(err, link.ErrNotSupported) {
+		// Kernel doesn't support bpf_link, clean up any other potential progs.
+		if err := searchAndDetach(name, cgroupRoot, prog); err != nil {
+			return fmt.Errorf("guarantee exclusivity for program %s: %w", name, err)
+		}
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("pin program %s link at %s: %w", name, pinPath, err)
+	}
+
+	// Successfully created and pinned bpf_link.
+	return nil
+}
+
+// exclusiveCgroupAttach attaches prog after first detaching all preexisting
+// programs from prog's corresponding cgroup hook.
+//
+// The detach operation can fail if one of the preexisting programs is attached
+// through a bpf_link not managed by Cilium.
+func exclusiveCgroupAttach(prog *ebpf.Program, name, cgroupRoot string) (link.Link, error) {
+	if err := searchAndDetach(name, cgroupRoot, nil); err != nil {
+		return nil, fmt.Errorf("detach program: %w", err)
+	}
+
+	l, err := link.AttachCgroup(link.CgroupOptions{
+		Path:    cgroupRoot,
+		Attach:  attachTypes[name],
+		Program: prog,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("attach program %s: %w", name, err)
+	}
+
+	return l, nil
+}
+
+// detachCgroup detaches a program with the given name from cgroupRoot. Attempts
+// to open a pinned link with the given name from directory pinPath first,
+// falling back to PROG_DETACH if no pin is present.
+func detachCgroup(name, cgroupRoot, pinPath string) error {
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Unpin(); err != nil {
+			return fmt.Errorf("unpin link %s: %w", pin, err)
+		}
+		if err := l.Close(); err != nil {
+			return fmt.Errorf("close link %s: %w", name, err)
+		}
+		return nil
+	}
+
+	// No bpf_link pin found, search and detach prog_attach progs.
+	log.Debugf("no pinned link '%s', querying cgroup", pin)
+	err = searchAndDetach(name, cgroupRoot, nil)
+	// Treat detaching unsupported attach types as successful.
+	if errors.Is(err, link.ErrNotSupported) {
+		return nil
+	}
+	return err
+}
+
+// tryPin pins the link to maintain it after agent exit. l is closed after
+// a successful pin operation.
+//
+// Returns ErrNotSupported if l was attached using PROG_ATTACH.
+func tryPin(l link.Link, name, pinPath string) error {
+	err := l.Pin(filepath.Join(pinPath, name))
+	if errors.Is(err, link.ErrNotSupported) {
+		log.Debugf("program %s was attached using PROG_ATTACH", name)
+	}
+	if err != nil {
+		return fmt.Errorf("pin link: %w", err)
+	}
+
+	log.Debugf("program %s attached using bpf_link", name)
+	return l.Close()
```

Comment:
(This behaviour won't be necessary after re-spin) You've sort of figured it out judging by your latest comment, but looping back to this: https://github.com/cilium/cilium/pull/23557/commits/8265e1f525e72e15871c2076413815f73bb04ce1#r1140300865.

If a link can be pinned (remember, ebpf-go returns an opaque `link.Link`, we can't really tell what was done under the covers), that means it's a bpf_link, which means it can safely be closed after pinning.

If it cannot be pinned, that must mean PROG_ATTACH happened, and we *must not* call `Close()`, since that will immediately call `PROG_DETACH`.

Reply:
:+1: Given the semantics of the API, that makes sense.

From other discussions it sounds like you're looking at this as well, but IMO those semantics are confusing for a library user to figure out and call `Close()` at the right  time. (I didn't look at how the API describes when the caller should call `Close()` or the semantics of link liveness underneath, but I'm sure you'll take a fresh look)